### ISSUE TITLE
Clarification request for Rule 2.7

### ISSUE THREAD

**Issue by pavelrappo** [reactive-streams/reactive-streams-jvm#405]
> A `Subscriber` MUST ensure that all calls on its `Subscription` take place from the same thread or provide for respective external synchronization.

Given [rule 1.3](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/README.md#1.3), technically, which thread a subscriber's method is called in is out of this subscriber's control. As far as I understand most often `Subscription`'s methods are called from `Subscriber`'s methods synchronously. 

Thus, either all calls to `Subscription` should be properly synchronized all the time (but that's not what we probably want) or the rule should be rephrased.

(Sorry if it looks like nitpicking.)

**Comment by akarnokd** [comment]
Rule [3.3](https://github.com/reactive-streams/reactive-streams-jvm#3.3) mandates limited recursion. The external synchronization and limited recursion can be solved via the same constructs such as dedicated actors, atomic state machines and/or the queue-drain serialization approach. 

> Thus, either all calls to Subscription should be properly synchronized all the time

Depends on the platform. Think about JavaScript where there is only one thread and thus everything is a synchronous call.

> As far as I understand most often Subscription's methods are called from Subscriber's methods synchronously.

Not necessarily. For example, an intermediate Publisher may prefetch data from its source even if the end consumer hasn't requested anything. This end-consumer can now request data which triggers the `onNext` emission by the Publisher. After a certain threshold, the Publisher may decide to prefetch even more data from its upstream, thus you have practically a `request()` call triggering more `request()` from upstream.

**Comment by viktorklang** [comment]
@pavelrappo RE you comment on 1.3, see 3.1: https://github.com/reactive-streams/reactive-streams-jvm#3.1

**Comment by pavelrappo** [comment]
@viktorklang This "context" the rule is talking about is a bit hard to define. I'm sure this question has been asked many times before, so pardon for asking it again: 

    What is this "context" the rule is talking about? 

Is it invocations of `Subscriber`'s methods? Is it any place (method, class, etc.) which the subscriber is aware of and permits calls to its `Subscription` from? (The thread the call is made in is obviously important, but is rather an orthogonal characteristic. Or is it not?)

**Comment by pavelrappo** [comment]
@akarnokd 
> Depends on the platform. Think about JavaScript where there is only one thread and thus everything is a synchronous call.

I think it's fair to say we are talking about JVM and threads. Otherwise this spec might need to become more general than it is today.

Suppose I have a general `Publisher` I don't know nothing about, except that is produces the type of items my `Subscriber` consumes. The `Subscriber` works as follows: in `onSubscribe` it requests a single item. After the item has been delivered to `onNext`, the subscriber starts a timer, say, half an hour. When the time has been done, the subscribe request one more item, rinse repeat.
How should the "external synchronization" in this case look like?

**Comment by viktorklang** [comment]
@pavelrappo The delivery of the "tick" from the timer needs to be delivered to the Subscriber somehow, how is that signal interleaved with signal (onX) processing?

**Comment by pavelrappo** [comment]
@viktorklang Not sure I understood your question. What I am asking though is how to provide "external synchronization" in this (schematic) case:
```java

    public class MySubscriber<T> implements Subscriber<T> {

        Subscription subscription;

        @Override
        public void onSubscribe(Subscription s) {
            // ...
            (subscription = s).request(1);
        }

        @Override
        public void onNext(T item) {
            // ...
            new Thread(() -> {
                try {
                    TimeUnit.MINUTES.sleep(30);
                } catch (InterruptedException e) {
                    subscription.cancel();
                    return;
                }
                subscription.request(1);
            }).start();
        }
        // ...
    }
```

**Comment by viktorklang** [comment]
@pavelrappo Your example would be problematic, since you're not safely publishing the Subscription between the thread which executes the onNext-signal and the thread which sleeps, and back.


Have you had a look at the examples? https://github.com/reactive-streams/reactive-streams-jvm/tree/master/examples/src/main/java/org/reactivestreams/example/unicast

**Comment by akarnokd** [comment]
Actually, `Thread.start()` acts as a barrier so store to the `subscription` field in `onSubscribe` happens before the execution of the `Runnable` which then uses that field.

The example shows that `request()` can be called from a different thread than `onNext` executes, thus the usual reentrancy-avoiding logic, such as setting a flag at the beginning of the `request()` implementation and clearing it at the end is not enough.

Depending on what the `Publisher` does, the thread-safe and reentrant-safe practical implementation could be:

- Create a `Request` message and send it to the Actor behind the `Subscription` as Actors process their message queue in ordered fashion on a (dedicated or pool) thread.
- Atomically add up the request amount while considering the `Long.MAX_VALUE` cap and if the previous request amount was 0L, emit items. This works with source-like `Publisher`s where the availability of items is not transitive in some sense, such as in intermediate stages. [Example](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/examples/src/main/java/org/reactivestreams/example/unicast/RangePublisher.java#L116).
- Do the same atomic add-and-cap like before, but perform an additional atomic increment on a counter. When this counter changes from 0 to 1, begin the emission of items. [Example in a blog post](http://akarnokd.blogspot.hu/2017/09/java-9-flow-api-switching-threads.html).

In all three cases, the `Subscriber` can call `Subscription.request()` from any thread at any time and get the correct and thread-safe behavior from the `Publisher`/`Subscription` implementation.

**Comment by pavelrappo** [comment]
@viktorklang 1) Thanks for the link. I will have a look at it no doubt. 2) I can't see how subscription is not visible to the created thread. 

1. `Publisher` guarantees happens-before between calling `onSubscribe` and `onNext`.
2. As per [JMM](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html): An action that starts a thread synchronizes-with the first action in the thread it starts.
**and**  If an action x synchronizes-with a following action y, then we also have hb(x, y) **and** If hb(x, y) and hb(y, z), then hb(x, z)
3. `onSubscribe` _happen-before_ `Thread.start`

Therefore, actions in `onSubscribe` _happen-before_ any action in a newly created thread.

**Comment by viktorklang** [comment]
@pavelrappo Yes, the barrier *to* the new thread is fine, but there is no barrier *back*. (You cannot assume that the Publisher is synchronous.

**Comment by viktorklang** [comment]
@akarnokd In your example, if I understand you correctly, assuming that there are no other, non-volatile fields involved, the addAndGet would provide the necessary barriers both to-and-from. BUT: publishing the Subscription still needs to be done safely, and concurrent signalling is still bad.

**Comment by pavelrappo** [comment]
@viktorklang I guess that's what my question was about. What external synchronization I should use to ensure `Subscription`'s methods are called in a thread-safe manner?

**Comment by viktorklang** [comment]
@pavelrappo Depends on how you implement your Subscriber. Is it a synchronous Subscriber? Is it async? If it is async, how is its execution controlled? Can your timer be hooked into that execution?

I think it makes sense for you to have a look at the examples, and see if that helps a bit or not. :)

**Comment by viktorklang** [comment]
Is everything addressed here?

**Comment by viktorklang** [comment]
@pavelrappo Ping

**Comment by pavelrappo** [comment]
@viktorklang yes, thank you

**Comment by viktorklang** [comment]
@pavelrappo perfect, thank you.

**Comment by thekalinga** [comment]
@viktorklang Does this rule exists to guarantee the Publisher that he sees state modifications that might be introduced inside `Subscription` object due to `request(long)`/`cancel`, which are further required for the `Publisher` to fulfil its requirements such as

1. Sending that many more `onNext`s corresponding to `request` &
2. Eventual cancellation of the subscription & thus release subscriber ref from publisher?

**Comment by viktorklang** [comment]
@thekalinga Thanks for you question, I hope I can answer it!

If all Subscriber implementations need to employ coordination for all their operations, it introduces overhead for everyone. Reactive Streams is not an end-user API, and it is not expected that developers implement any of the interfaces without understanding the spec or verifying the implementation using the TCK. See this [example](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/examples/src/main/java/org/reactivestreams/example/unicast/AsyncSubscriber.java) to get an overview of which concerns need to be addressed in the normal case for an async Subscriber implementation.

Cheers,
√

**Comment by thekalinga** [comment]
@viktorklang Thanks for your response. I've downloaded the source. Will go over all the implementation before I can ask any (semi) educated questions

**Comment by thekalinga** [comment]
@viktorklang One things comes to mind (posting here so that I might not forget it)

> Yes, the barrier to the new thread is fine, but there is no barrier back. (You cannot assume that the Publisher is synchronous.

Can you please clarify what does "barrier back" mean. What does the publisher "not see" if there is no external synchronization from subscriber as the publisher is expected to send in a thread safe manner anyways

**Comment by viktorklang** [comment]
@thekalinga If you think about it in a fork+join manner, there is a fork (the new thread) but there is no join phase.

**Comment by thekalinga** [comment]
@viktorklang But the underlying reason is still not explained. What happens when you don't join & why is join even required? That's what I'm trying to find as the rule says we need to synchronise some how, but why synchronise in the first place on the subscriber side?

**Comment by viktorklang** [comment]
@thekalinga The Subscriber needs to do coordination if it runs asynchronously, and since concurrent signalling is forbidden, if it exposes the Subscription it needs to ensure that no concurrent signalling takes place.

**Comment by thekalinga** [comment]
Looks like I got the answer. Will post here for others who might have the same query as me

> From Spec explanation of 2.7:The intent of this rule is to establish that external synchronization must be added if a Subscriber will be using a Subscription concurrently by two or more threads.

This implies that its required because of state management inside `Subscription`

My question was "If above is true, shouldn't this be the responsibility of `Publisher` to make sure that whatever the `Subscription` implementation he is using is thread safe when `request` & `cancel` are invoked? Why should `Publisher` expect his object `Subscription`'s state be guarded by ``Subscriber`?"

After thinking about it for a while, I see that most subscribers respond (`request`/`cancel`) within the same call, so any thread safety code `Publisher` might put inside `Subscription` to make `Subscriber` life simple  might be a premature optimisation for most subscribers & will degrade their performance

In the rare case that `Subscriber` does request more asynchronously via `request`, if the `Subscriber` cooperates with producer in making sure that the Subscription changes are seen by the `Publisher`, `Publisher` can do his job of signalling correctly

**Comment by thekalinga** [comment]
Not sure whether its correct. Can you please confirm?

**Comment by thekalinga** [comment]
> The Subscriber needs to do coordination if it runs asynchronously, and since concurrent signalling is forbidden, if it exposes the Subscription it needs to ensure that no concurrent signalling takes place.

Why does `request` & `cancel` should not be performed concurrently? What are we trying to achieve by doing this?

**Comment by viktorklang** [comment]
@thekalinga Let's take a step back—what problem are you looking to solve?

**Comment by thekalinga** [comment]
@viktorklang I'm trying to understand the reason for the existence of this rule in reactive streams spec. I'm creating a training material for each of the rules of reactive streams spec

PS: Thanks for your time

**Comment by viktorklang** [comment]
@thekalinga Ok. :) My hope would be that the Intent sections for these rule answer why they exist. For 2.7 specifically, it is about setting the appropriate expectations for Subscriber implementations. I've found a good exercise to be to imagine that the rule doesn't exist, and what would happen if code didn't take it into consideration.

**Comment by thekalinga** [comment]
@viktorklang Is the above explanation https://github.com/reactive-streams/reactive-streams-jvm/issues/405#issuecomment-414318349 is on the mark/did I go wrong somewhere?

**Comment by thekalinga** [comment]
The current explanation is too terse & does not give any underlying reasons. It just says "do this" & does not explain clearly why one should?

If there is an explanation of what happens if this is not followed, this will give people clear justification of why this rule exists. Most of the rules of reactive streams are easy to follow & obvious to anyone with basic understanding of java threading, but rules like this are not that obvious

**Comment by viktorklang** [comment]
@thekalinga Perhaps the easiest answer is that the rule exists to prevent un-coordinated calls to `request`, and to make sure that the Subscription is safely published between threads.

@reactive-streams/contributors Anyone have a better phrased explanation?

**Comment by thekalinga** [comment]
@viktorklang Since a hot `Publisher` can have many `Subscriber`s, `Publisher` is expected to guard its `request`/`cancel` AFAIK. And concurrent calls are the norm for `request`. So I'm not sure whether its the full explanation

**Comment by viktorklang** [comment]
@thekalinga Any Publisher can have many Subscribers, but they all get their own Subscription. It's not whether the Subscriber will run on another thread, it is about if the Subscriber uses the Subscription on many threads *concurrently to each other* not *concurrently with the Publisher*.

**Comment by thekalinga** [comment]
@viktorklang Yes you are right that every subscriber will have its own subscription

>  It's not whether the Subscriber will run on another thread, it is about if the Subscriber uses the Subscription on many threads concurrently to each other not concurrently with the Publisher.

In most cases the sideeffects of `Subscription` methods would be made inside the `Publisher` to track demand, even though user would be invoking methods on `Subscription`

But I am not sure why the above needs to be done by `Subscriber`. Asking `Subscriber` to do this job breaks Single responsibility principle & encapsulation. If the passed `Subscription`/`Publisher` encapsulates thread safety inside itself, there is no need for `Subscriber` to worry about whether the passed in `Subscription` is thread safe or not. If the `Subscription`/`Publisher` pair is already takes care of thread safety issues (in RxJava it does as per @akarnokd), its unnecessary to do one more round of thread safety checks in `Subscriber` & even unreasonable for us to expect `Subscriber` to deal with the implementation details because `Subscription` did not take care of its own thread safety

I had a conversation with @akarnokd on twitter. Adding that here as a reference

https://twitter.com/1tontech/status/1031548972572110858

**Comment by viktorklang** [comment]
@thekalinga 

>But I am not sure why the above needs to be done by Subscriber. 

I believe I have already explained *why* coordination needs to be done by the Subscriber if the 
Subscriber implementation has decided to use possibly different threads of execution at different, or possibly the same, time. I'm not sure how I can say it in a different way? :/

>Asking Subscriber to do this job breaks Single responsibility principle & encapsulation

Absolutely not. *IF* the Subscriber decides to use multiple threads then *of course* the Subscriber is required to coordinate access to its resources. I hope we can agree here.

> (Please note that any rule can broken but only with a good justification like say performance impact)

???

>. If the passed Subscription/Publisher encapsulates thread safety inside itself (which I believe it should), there is no need for Subscriber to worry about whether the passed in Subscription is thread safe or not

It is a bit of a truism to say that coordination doesn't need to be done if coordination is already done.

I would hope that it is a bit of a non-issue, as don't you agree that if someone implements a multi-threaded Subscriber, they had better understand how to do so safely?

>If the Subscription/Publisher pair has already taken care of thread safety issues (in RxJava Subscription does take care of this aspect as per @akarnokd), its unnecessary to do one more round of thread safety checks in Subscriber & even unreasonable for us to expect Subscriber to blindly deal with the thread safety because Subscription might not have taken care of its own thread safety

See above.

>The thread safety responsibility of Subscription methods are better encapsulated in Subscription itself instead of pushing them to Subscriber

By that logic, should all collections in Java be thread safe?

>I had a conversation with @akarnokd on twitter. Adding that here as a reference. Can you please go over it.

Better to ask @akarnokd to reply here

**Comment by viktorklang** [comment]
@thekalinga Please let me know if there's anything I can elaborate on which would be of help, I'm not sure I can add much else to this specific conversation. Some other @reactive-streams/contributors might have a different perspective than I. Thanks for raising the question!

**Comment by thekalinga** [comment]
@viktorklang Thanks for your response

> By that logic, should all collections in Java be thread safe?

No because `Subscriber` is not initialising `Subscription` & is clueless about how demand is managed inside `Publisher` when he makes calls on `Subscription`.  Since `Publisher` is creating `Subscription`, he knows what kind of implementation is being used for `Subscription` & should be in a position to make a judgement as to how to protect from multi threaded calls

For eg., If I create an object, I am in a better position to handle thread safety efficiently than the users of my object. I assume the reason of non thread safe collections in java is performance

Everything is boiling down to should own the responsibility. Tho `Subscriber` owns `Subscription` (one to one mapping), the implementation used for `Subscription` is known only by `Publisher` as he creates it and passes down, so is in a better position to guard than `Subscriber`

> I would hope that it is a bit of a non-issue, as don't you agree that if someone implements a multi-threaded Subscriber, they had better understand how to do so safely?

Sure, he is is. But do u agree that in multiple subscriber case (hot source), `Publisher` must take care of thread safety even if `Subscriber` has ensured thread safety before he called `request`, as simultaneous `request`s from multiple `Subscription`s from different threads can still make `Publisher` unsafe unless thread safety is ensured inside `Publisher`? If so, we might aswell not force `Subscriber` to take care of multi threaded issues, as we will have double synchronisation, one at Subscription & another at Publisher level aswell

In case of *cold publisher* case, it will give some performance edge if async subscribers (those that call `Subsciption` asynchronously) is forced to take care of multil threading, i.e those subscribers that are using `Subscription` synchronously will not pay any synchronisation penalty.

If `Subscriber` is forced to own the synchronisation & Publisher too is not relieved of synchronisation issue (hot publisher case), why not push this responsibility to `Subscriber`/`Publisher` combination, who knows how to deal concurrency in the most effective way because they know the implementation details, where as `Subscriber` is not

I only see that the *cold publisher* case gives some performance benefits to justify `Subscriber` owning this responsibility, else `Publisher` should own this responsibility (which is what I mentioned here aswell https://github.com/reactive-streams/reactive-streams-jvm/issues/405#issuecomment-414318349)

Please let me know your views

**Comment by thekalinga** [comment]
Updated my previous comment as I found it somewhat incorrect. Now it reflects my views correctly

**Comment by DougLea** [comment]
I'm a little confused by some parts of this exchange. But maybe this helps: There is a classic race condition on any variable representing demand: the subscriber increases, the publisher decreases. If, in some special cases,  you can somehow arrange that this race does not occur, you can relax synchronization. But in general, you cannot do so while still conforming to spec.

**Comment by viktorklang** [comment]
@thekalinga Thanks for updating the comment, I hope I'm answering your questions as it is likely that this response will be my last in this conversation as I am not sure I have anything more to add here, I'm afraid. :(

>>By that logic, should all collections in Java be thread safe?

>No because Subscriber is not initialising Subscription & is clueless about how demand is managed inside Publisher when he makes calls on Subscription. 

Imagine the following scenario:

<T> void onSubscribe(java.util.List<T> list) {
   // Would you assume that `list` is thread safe here?
}

>Since Publisher is creating Subscription, he knows what kind of implementation is being used for Subscription & should be in a position to make a judgement as to how to protect from multi threaded calls and also do it efficiently

Since a Subscription represents a 1-to-1 relationship, it does not need to assume that it is going to be invoked in parallel by its Subscriber, without sufficient coordination. This is a classic situation of Single-Producer-Single-Consumer vs Multiple-Producer-Single-Consumer. These different modes have different performance (and scalability) implications.

>For eg., If I create an object, I am in a better position to handle thread safety efficiently than the users of my object when the user makes a call on any of my object. I assume the reason of non thread safe collections in java is performance, but who ever owns them should take care of the responsibility of managing the thread safety rather than asking all its users to worry about it. So all Publishers operating in multi threaded environment should be aware that his methods can be invoked simultaneously from multiple threads so should guard against that scenario

If your position is—and I don't know that it is—that all classes should be thread-safe, then I'm certain that this is not the place for that conversation, and I hope you agree with me there.

>Everything is boiling down to who owns the responsibility. Tho there is one to one mapping between Subscription & Subscriber, the implementation used for Subscription is known only by Publisher as he creates it and passes down, so is in a better position to guard than Subscriber

What I know is that nothing is in a better position to deal with the thread-safety of a Subscriber which chooses to implement its own functionality using multithreading than in that very Subscriber.

>>I would hope that it is a bit of a non-issue, as don't you agree that if someone implements a multi-threaded Subscriber, they had better understand how to do so safely?

>Sure, he is is. But do u agree that in multiple subscriber case (hot source), Publisher must take care of thread safety even if Subscriber has ensured thread safety before he called request, as simultaneous requests from multiple Subscriptions from different threads can still make Publisher unsafe unless thread safety is ensured inside Publisher? If so, we might aswell NOT force Subscriber to take care of multi threaded issues, as we will have double synchronisation, one at Subscriber & another at Publisher level aswell

I'm not sure that we understand each other, so let me try to clarify my perspective: We are talking about the implementation of Subscriber. In the case of synchronous subscribers, a hot Publisher *still* has to manage coordination (thread safety and semantics) for possibly many Subscribers, since they could be subscribed from different threads. What 2.7 is about is when the implementation of Subscriber decides it's going to use multi-threading *internally* (or call out to some Executor or otherwise). So what the rule says is that if you are going to implement a Subscriber which uses multi-threading, you need to ensure safe publication of the Subscription and to ensure that there's a happens-before relationship between request-invocations.

>In case of cold publisher case, it will give some performance edge if async subscribers (those that call Subsciption asynchronously) is forced to take care of multil threading. This way those subscribers that are using Subscription synchronously will not pay any synchronisation penalty.

>If Subscriber is forced to own the synchronisation & Publisher too is not relieved of synchronisation issue (hot publisher case), why not push this responsibility to Subscription/Publisher combination, who knows how to deal concurrency in the most effective way because they know the implementation details, where as Subscriber is not

Many Subscribers are actually also Publishers, since few streams are a single Publisher-Subscriber pair.

>I only see that the cold publisher case gives some performance benefits to justify Subscriber owning this responsibility, else Publisher should own this responsibility (which is what I mentioned here aswell #405 (comment))

To my knowledge at this point: changing the spec to require that all Subscriptions are thread-safe under all usage patterns would be a breaking change, and breaking the compatibility of the spec would require extraordinary reasons. I'm sure you agree that the burden of evidence has to be on the person who proposes such a breaking change.

Cheers,
√

**Comment by thekalinga** [comment]
@DougLea @viktorklang

Thanks for your effort in helping me understand this. I tried my level best to explain what I'm trying to ask in this reply






To make things concrete, I wrote the following mock code. In this example we have multiple `Publisher`s & `Subscriber`s (Please note that this is pseudo code & is not spec complaint but only written to demonstrate the point I'm trying to make)

As you can see `Safe27NonCompliantSubscriber` violates 2.7, yet safe to use with `SafePublisher`. Also, `Unsafe27CompliantSubscriber` complies with 2.7, yet an unsafe `Subscriber`

```java
// (mostly) works correctly with any subscriber even when demands are made simultaneously
// by multiple subscribers (i.e both as hot & cold publisher). Also well encapsulated
// `Subscription`/`Publisher` pair who manage thier internal state without any external help from `Subscriber`
class SafePublisher<T> implements Publisher<T> {
  AtomicLong unfilledDemand = new AtomicLong();

  @Override
  public void subscribe(Subscriber<? super T> subscriber) {
    subscriber.onSubscribe(new SafeSubscription());
  }

  class SafeSubscription implements Subscription {
    @Override
    public void request(long n) {
      for (;;) {
        long prevDemand = unfilledDemand.get();
        if (unfilledDemand.compareAndSet(prevDemand, prevDemand + 1)) {
          return;
        }
      }
    }

    @Override
    public void cancel() {
      unfilledDemand.set(0);
    }
  }

  void fulfillDemandSomehow() {
    // fulfills demand
  }
}

// rarely works correctly
class UnsafePublisher<T> implements Publisher<T> {
  long unfilledDemand;

  @Override
  public void subscribe(Subscriber<? super T> subscriber) {
    subscriber.onSubscribe(new UnsafeSubscription());
  }

  class UnsafeSubscription implements Subscription {
    @Override
    public void request(long n) {
      unfilledDemand += n;
    }

    @Override
    public void cancel() {
      unfilledDemand = 0;
    }
  }

  void fulfillDemandSomehow() {
    // fulfills unpredicatble amount of demand
  }
}

// 2.7 complaint safe subscriber (I hope its complaint :) )
class Safe27CompliantSubscriber<T> implements Subscriber<T> {
  Subscription s;
  AtomicInteger itemCount = new AtomicInteger();

  @Override
  public void onSubscribe(Subscription s) {
    this.s = s;
    new Thread(() -> {
      itemCount.addAndGet(1); // happens-before contract
      s.request(1);
    }).start();
  }

  @Override
  public void onNext(T t) {
    new Thread(() -> {
      int total = itemCount.addAndGet(1); // happens-before contract
      if (total < 10) {
        s.request(1);
      }
    }).start();
  }

  @Override
  public void onComplete() { }

  @Override
  public void onError(Throwable t) { }
}

// 2.7 complaint, but unsafe subscriber
class Unsafe27CompliantSubscriber<T> implements Subscriber<T> {
  PrivateState state;

  @Override
  public void onSubscribe(Subscription s) {
    // 2.7 complaint as we are only making calls on `Subscription` from within `Publisher` thread
    s.request(MAX_VALUE);
  }

  @Override
  public void onNext(T t) {
    new Thread(() -> {
      // No call to `Subscription` from non `Publisher` thread, yet have to guard concurrant modifications to `state` to work correctly
      // If not guarded properly bad things will happen
      doUnsafeSideEffectsToPrivateState(t);
    }).start();
  }

  @Override
  public void onComplete() { }

  @Override
  public void onError(Throwable t) { }
}

// 2.7 non complaint, but safe subscriber, takes care of its internal state (as it has side effect free actions)
class Safe27NonCompliantSubscriber<T> implements Subscriber<T> {
  Subscription s;

  @Override
  public void onSubscribe(Subscription s) {
    this.s = s;
    new Thread(() -> {
      s.request(1);
    }).start();
  }

  @Override
  public void onNext(T t) {
    new Thread(() -> {
      doSideEffectFreeAction(t);
      s.request(1);
    }).start();
  }

  @Override
  public void onComplete() { }

  @Override
  public void onError(Throwable t) { }
}

// 2.7 non complaint, unsafe subscriber
class UnSafe27NonCompliantSubscriber<T> implements Subscriber<T> {
  Subscription s;
  PrivateState state;

  @Override
  public void onSubscribe(Subscription s) {
    this.s = s;
    new Thread(() -> {
      s.request(1);
    }).start();
  }

  @Override
  public void onNext(T t) {
    new Thread(() -> {
      //Have to guard concurrant modifications to `state` to work correctly. If not guarded properly bad things will happen
      doUnsafeSideEffectsToPrivateState(t);
      s.request(1);
    }).start();
  }

  @Override
  public void onComplete() { }

  @Override
  public void onError(Throwable t) { }
}

class Entrypoint {
  public static void main(String[] args) {
    SafePublisher<Integer> safePublisher = new SafePublisher<Integer>();
    safePublisher.subscribe(new Safe27CompliantSubscriber<>());
    safePublisher.subscribe(new Unsafe27CompliantSubscriber<>());
    safePublisher.subscribe(new Safe27NonCompliantSubscriber<>());
    safePublisher.subscribe(new UnSafe27NonCompliantSubscriber<>());

    UnsafePublisher<Integer> unsafePublisher = new UnsafePublisher<Integer>();
    unsafePublisher.subscribe(new Safe27CompliantSubscriber<>());
    unsafePublisher.subscribe(new Unsafe27CompliantSubscriber<>());
    unsafePublisher.subscribe(new Safe27NonCompliantSubscriber<>());
    unsafePublisher.subscribe(new UnSafe27NonCompliantSubscriber<>());
  }
}
```

For all the above examples, which of the `Publisher`/`Subscription` combination is desirable, i.e which of them encapsulate their state well. I believe `SafePublisher`

This is what the spec says

> 2.7 Rule:	A `Subscriber` MUST ensure that all calls on its `Subscription` take place from the same thread or provide for respective external synchronization.

> Explanation: The intent of this rule is to establish that external synchronization must be added if a `Subscriber` will be using a `Subscription` concurrently by two or more threads.

My question has been, "how does its matter which thread the `Subscriber` invokes the `Subscription` methods so long as `Publisher` has ensured thread safety on its own for its internal state changes and ensured *happens-before* between multiple downstream signals by ensuring thread safety during `request` & `cancel`?"

If `Subscriber` has not taken care of `happens-before` relationship, its should not affect `Publisher` contracts in anyway, it should only affect the correctness of `Subscriber`. `Subscriber` should take care of his own internal state management & so is `Publisher` when it comes to `Subscription`

But the rule states this `Subscriber` synchronization has something to do with `Subscription`, but I am not sure if it is/it should be (unless ofcourse there is a performance cost involved if `Publisher` does this), as `Subscription` can be well encapsulated by `Publisher` (I hope the above mock code demonstrates this)

@viktorklang

> you need to ensure safe publication of the Subscription and to ensure that there's a happens-before relationship between request-invocations

I believe `happens-before` relationship inside `Subscriber` is for making sure that the `Subscriber` state is intact from its own point of view between multiple calls from `Publisher` & has nothing to do with *happens-before* relationship inside `Subscription`/`Publisher` (or) thier safety . Is that correct?

If correct, what does "safe publication of `Subscription`" means here (also inside the 2.7 rule)? If not, I'm not sure why it has anything to do with `Publisher`/`Subscription` safety

I'm sure I'm missing something obvious (may be my lack of clear understanding of multi threading or something) (or) I'm not seeing that's obvious in front of my eyes. Appreciate if anyone can clarify

If anyone can give a concrete example on how this rule helps as I believe I have provided counter examples for this rule

**Comment by DougLea** [comment]
I agree that "safely concurrent" and "obeys all rules" are NOT synonymous. And that it is possible to follow 2.7 and still construct an unsafe component. Do you have any rewording in mind that would help here?

**Comment by thekalinga** [comment]
@DougLea Do u agree with this statement

"`Subscriber` synchronisation is only for `Subscriber` state safety between multiple signals from `Publisher` & has nothing to do with `Subscription`/`Publisher` state. Also, the thread from which `Subscription` methods are invoked should not affect the correctness of `Publisher` behaviour"

If this can be agreed upon, then the changes to the definition can be worked out

Please let me know if that's your understanding too/I'm wrong anywhere

**Comment by DougLea** [comment]
To rephrase: I agree that subscriber synchronization is not necessarily required to maintain protocol state correctness, but may be necessary to maintain subscriber state correctness.

**Comment by viktorklang** [comment]
I believe that it is perfectly valid according to the spec to create a Subscription which tracks a `delivered` `long` and a `demanded` `long`.

Such a Subscription would work perfectly fine as long as there is an external guarantee that `request` happens-before the next `request`.

And to reiterate, suddenly mandating that all Subscriptions be thread-safe in all interaction patterns would not be forwards-compatible (new Subscribers calling old Publishers).

**Comment by thekalinga** [comment]
Thank you both for your responses

@DougLea I too concur with you

@viktorklang 

> Such a Subscription would work perfectly fine as long as there is an external guarantee that request happens-before the next request.

The tricky part is this section of your statement *external guarantee that request happens-before the next request*

As a contract, do you agree that an unsafe `Subscriber` should not break the functionality of `Publisher`/`Subscription`? Especially when we know in advance that end users like me (`Subscriber`s) struggle to correctly maintain their own state, why should any upstream component (`Publisher`s & `Processor`s) expect `Subscriber` to take of the guarantees for upstream state as well in addition to `Subscriber`s own state?

Unless `Subscriber` needs to do this for some performance gain (which again is not yet demonstrated), as of now, I dont see any serious justification for 2.7

> A Subscriber MUST ensure that all calls on its Subscription take place from the same thread or provide for respective external synchronization.

Other than the forwards compatibility, as a principle, do u disagree @DougLea's understanding?

Can you please clarify

**Comment by viktorklang** [comment]
@thekalinga 

Reactive Streams is a akin to a SPI, and end-users are not expected to implement the interfaces themselves—unless they know they can adhere to the spec and verifies the implementation with the TCK— instead they should be using implementations such as, but not limited to, Akka Streams, Reactor, RxJava etc. The benefit comes from being able to combine such implementations to get correct async, non-blocking, back-pressured streams.

>As a contract, do you agree that an unsafe Subscriber should not break the functionality of Publisher/Subscription? Especially when we know in advance that end users like me (Subscribers) struggle to correctly maintain their own state, why should any upstream component (Publishers & Processors) expect Subscriber to take of the guarantees for upstream state as well in addition to Subscribers own state?

So, to be clear, we are talking about the situation where someone decides to create a multithreaded Subscriber. Conceptually, the Subscription is a Single-Producer-Single-Consumer item, if you decide to use it differently then it is up to you to ensure correctness.

Whether or not a Subscriber should be able to "break" a Subscription by not adhering to the spec—well the spec is there for a reason, and there are plenty of cases where a faulty, or even Byzantine, Subscriber implementation can cause all kinds of problems. (And, the Subscription becomes a part of the Subscriber's state, and it needs to deal with memory visibility for its state.)

I'm afraid that unless anything new (substantial) is added to this conversation, I have nothing else to add.

Cheers,
√

**Comment by thekalinga** [comment]
@viktorklang 

Agree with you for the SPI & implementation parts. Sure end users like me rarely implement our own `Publisher`/`Processor`, but needs to implement `Subscriber`s for any non trivial activity

Even if the above one is granted, I dont think it justifies current form of 2.7 the way its framed, linking `Subscriber` to take care of `Subscription` state. It clearly breaks *Single Responsibility Principle*

`Subscription` is `Publisher`s agent given to `Subscriber` to talk back to `Publisher`. So, the state responsibility to ensure state consistency should lie with the implementors of `Subscription`, not `Subscriber`

Can someone else please elaborate justification for 2.7 as I dont see any as of now (other than the fact that some `Publisher`s *might* have little performance advantage if `Subscriber` cooperates this way)

**Comment by rkuhn** [comment]
@thekalinga You are right in that there is nothing forcing the spec to contain rule 2.7, that rule is a design choice (just like all the others). The justification is exactly what you already confirmed, namely that Reactive Streams places the burden of synchronization on the caller of each callback, be that the Subscriber methods or the Subscription methods (cancel is a bit different since it is best effort anyway). With this scheme it becomes practical to implement very lightweight (single-thread) Publishers and Subscribers while still maintaining full compatibility with heavyweight (multi-thread) implementations. The choice is thus more social than technical—just like the origin of the whole standard.

BTW: can you elaborate on the need to implement `Subscriber` for any non-trivial activity? RxJava, Reactor, Akka Streams _et al_ all contain very powerful operators that allow the recipient of a stream to do quite non-trivial things without every needing to implement a Subscriber by hand.

**Comment by thekalinga** [comment]
@rkuhn

Thanks for your response. I agree with you only partially. Lemme justify my reasoning

I assume by `callback`s you are referring to method invocations on `Subscriber` & `Subscription` interfaces

Please correct me if I'm wrong in any of my analysis below

We have rule `1.3` that refers to contract for invocations on `Subscriber` & rule `2.7` that refers to contract for invocations on `Subscription`

> Rule 1.3: `onSubscribe`, `onNext`, `onError` and `onComplete` signaled to a Subscriber MUST be signaled in a thread-safemanner—and if performed by multiple threads—use external synchronization.

> Rule 2.7 A `Subscriber` MUST ensure that all calls on its `Subscription` take place from the same thread or provide for respective external synchronization.

My assumption is that reason for the existence of rule `1.3` is to make the life easy for `Subscriber`s who are usually not that well versed in multi threading (people like me) when compared to the developers of `Publisher`s & `Processor`s. With rule `1.3`, people like me gets a simple guarantee. Irrespective of how the upstream is implemented, what ever the state changes I had in my `Subscriber` object at the start of previous upstream signal will still be seen by the next signal from upstream

So from my point of view (lay developer) its a justified development pain (& performance cost due to additional synchornization) for framework developers to implement rule `1.3`. They are making my life simple with this

But rule `2.7` inverts the responsibility & expects the lay developer to not only worry about the state in his own `Subscriber` but also about the `Subscription` (and `Publisher`) state. So from this point of view, linking `Subscriber` to `Subscription` state is not justified (unless there is really sound reasoning, which I have not yet seen)

> BTW: can you elaborate on the need to implement Subscriber for any non-trivial activity? RxJava, Reactor, Akka Streams et al all contain very powerful operators that allow the recipient of a stream to do quite non-trivial things without every needing to implement a Subscriber by hand.

For eg., To achieve back pressure in a `Subscriber` who does simple background job (Lemme use Reactor as an example)

I cant use `Flux.subscribe(onNext, onError?, onComplete?)` as it requests `Long.MAX_VALUE`. The other alternative method `subscribe(onSubscribe, onNext, onError?, onComplete?)` needs me store Subscription in the outer scope

But the implementation of this is a direct pass thru to upstream (rule `2.7` is not taken care of I believe must be for performance reason)

https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/Flux.java#L7688-L7696

https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/LambdaSubscriber.java#L80

So I write my own `Subscriber` which using `subscribe(Subscriber)` and reactor wraps in `StrictSubscriber`

https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/Flux.java#L7700

https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/Operators.java#L1052

If I'm not mistaken, even `StrictSubscriber` does not take care of rule  `2.7` automatically (I assume for again for performance reasons). So end users like me are forced to do follow rule `2.7` because upstream publisher has not added thread safety guarantees inside his `Subscription` implementation

https://github.com/reactor/reactor-core/blob/5fa95df3f433cf388db65839557cda9dc548b629/reactor-core/src/main/java/reactor/core/publisher/StrictSubscriber.java#L128-L150

PS: Please note that I have limited understanding of reactor & I dont know if there are better ways to achieve the same back pressure

**Comment by rkuhn** [comment]
@thekalinga Thanks for elaborating on your understanding of the roles of Publisher and Subscriber, this is the crucial point. Your desire to question rule 2.7 stems from the view that Subscriber is meant to be implemented by end-users while Publisher (and thus Subscription) is not. This is not correct! Subscriber is intended for the very same audience as Publisher, namely for library authors exclusively.

This basic assumption lies at the bottom of this thread, please confirm that we can close it: Reactive Streams will not be changed to target end users as far as I am aware.

PS: my understanding of Reactor is also limited, in Akka Streams you would perform your extensive background job wrapped in a CompletionStage and wire that into your stream using `.mapAsync` with parallelism 1.

**Comment by thekalinga** [comment]
@rkuhn 

> This is not correct! Subscriber is intended for the very same audience as Publisher, namely for library authors exclusively.

Ok. End users like me will struggle with this rule then

> This basic assumption lies at the bottom of this thread, please confirm that we can close it: Reactive Streams will not be changed to target end users as far as I am aware.

If its set in stone, it can be closed

>  in Akka Streams you would perform your extensive background job wrapped in a CompletionStage and wire that into your stream using .mapAsync with parallelism 1.

I'm completely clueless about Akka. Can you please share pseudo code in Akka (may its doable in reactor too similarly)

**Comment by thekalinga** [comment]
@rkuhn Now that you mentioned its only for library developers, can you briefly mention whats the justification from the point of view of `1.3`  & `2.7` (as my justifications are obviously incorrect)?
